File: 백준/Gold/8980. 택배/택배.py
N, C = map(int, input().split())
M = int(input())
boxes = []

for _ in range(M):
    a, b, c = map(int, input().split())
    boxes.append((a, b, c))

boxes.sort(key=lambda x : x[1])

village = [C] * (N+1) # 각 마을에서 싣을 수 있는 택배의 개수
ans = 0
for s, e, c in boxes:
    tmp = C
    for i in range(s, e): # 출발지부터 도착지까지
        tmp = min(tmp, village[i]) # 옮길 수 있는 것과 상한선 중 작은 값
    tmp = min(tmp, c) # 옮길 수 있는 상자 개수랑 실제 박스 개수
    for k in range(s, e):
        village[k] -= tmp
    ans += tmp
print(ans)

File: 백준/Gold/8980. 택배/택배.java
import java.util.*;
import java.io.*;
public class Main {
    public static class Box implements Comparable<Box> {
        int s, e, cnt;

        public Box(int s, int e, int cnt) {
            this.s = s;
            this.e = e;
            this.cnt = cnt;
        }

        public int compareTo(Box o) {
            if (this.e == o.e) {
                return this.s - o.s;
            }
            return this.e - o.e;
        }
    }

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int C = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(br.readLine());

        ArrayList<Box> box = new ArrayList<>();

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());

            int s = Integer.parseInt(st.nextToken());
            int e = Integer.parseInt(st.nextToken());
            int cnt = Integer.parseInt(st.nextToken());

            box.add(new Box(s, e, cnt));
        }

        Collections.sort(box);

        int[] village = new int[N];
        Arrays.fill(village, C);

        int ans = 0;
        for (int i = 0; i < M; i++) {
            Box b = box.get(i);

            // 출발지에서 도착지까지 옮길 수 있는 최대 무게
            int min = C;
            for (int j = b.s; j < b.e; j++) {
                min = Math.min(min, village[j]);
            }

            int tmp = Math.min(b.cnt, min);

            for (int j = b.s; j < b.e; j++) {
                village[j] -= tmp;
            }
            ans += tmp;
        }

        System.out.print(ans);

    }
}
